README.md
popclient
=========

This repository is the "official" final resting place of the original 
`popclient` source code written by Carl Harris around 1993. `popclient` 
is a very small and simple POP3/POP2 email client that was developed 
in a time before POP and subsequently IMAP became the de-facto standards 
for e-mail retrieval on today's Internet. 

It is preserved here only for the sake of posterity, and almost certainly
shouldn't be used today. You might have read about `popclient` 
in Eric S Raymond's [The Cathedral and the Bazaar](http://www.catb.org/esr/writings/cathedral-bazaar/cathedral-bazaar/) or in Nick Christenson's 
[Sendmail Performance Tuning](https://books.google.com/books/about/Sendmail_Performance_Tuning.html) and perhaps have been curious to see (or even try to
build) the source. You might also be be interested in reviewing the original [man page](https://soulwing.github.io/popclient/popclient.html).

Around 1996 `popclient` was radically transformed by Eric Raymond into
the `fetchmail` utility. In 2004, [Rob Funk](https://sourceforge.net/u/robfunk/profile/) and others took responsibility for the ongoing maintenance and of
fetchmail, and as of 2013 it was still being actively maintained at
[www.fetchmail.info](http://www.fetchmail.info). If you're looking for 
command-line interface for POP, IMAP, etc, Fetchmail is for you.


CHANGES

popclient
Copyright (c) 1993,1994  Carl Harris, Jr.  <ceharris@vt.edu>

Version	Description
------- -----------
1.0- 	o  Original POP2 development version, term support, minor patches.
1.6	

1.7	o  Added MAIL environment variable support, System V Makefile changes.

2.1	o  First version to support POP3.
	o  First wide release to public at large.
	o  Began maintaining this file of changes.

2.1a	o  Fixed bug in verbose output in POP3 mode.
	o  Changed bzero() on the options structure to account for programmer 
           boneheadedness.
	o  Changed errorneous call in SockPrintf from SockPuts to SockWrite.
	o  Moved error report on socket errors in main().
	o  Removed -g -O0 from CFLAGS in Makefile.

2.1b	o  Changed erroneous fputs() to fprintf() in POP3_OK.

2.20	o  Changed argv[] pointers to point to innocuous junk so that a
	   password specified using -p will not be compromised by the
	   'ps' command.
	o  Radically changed system mailbox handling to use /bin/mail (or
	   equivalent MDA) to deliver user's mail to the mail spool.  This
	   not only obviates the need for MAIL environment variable support,
	   but also relieves the client of folder locking concerns.  Thanks
	   go to Tor Lillqvist <tml@tik.vtt.fi> for this suggestion.
	o  Set up to allow installer to specify the default protocol and the
	   new MDA (/bin/mail) support configuration in config.h.
	o  Typecast call to getpass() to get rid of gcc warning.
	o  Check for INADDR_NONE definition in socket.h

2.21	o  Changed inline function argument definitions in socket.c to easy
	   SunOS compile.
	o  Fixed mail delivery to system mail folder to use _login_ ID not
	   POP mail user ID.



pop2.c
/* Copyright 1993,1994 by Carl Harris, Jr.
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Carl Harris <ceharris@vt.edu>
 */


/***********************************************************************
  module:       pop2.c
  program:      popclient
  SCCS ID:	@(#)pop2.c	1.4  3/31/94
  programmer:   Carl Harris, ceharris@vt.edu
  date:         29 December 1993
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  POP2 client code.
 ***********************************************************************/
 
#include  <stdio.h>
#include  <sys/time.h>
#include  <unistd.h>
#include  <errno.h>

#include  "socket.h"
#include  "config.h"
#include  "popclient.h"

/* requisite, venerable SCCS ID */
static char sccs_id [] = "@(#)pop2.c	1.4\t3/31/94";

/* TCP port number for POP2 as defined by RFC 937 */
#define	  POP2_PORT	109

#ifndef NO_PROTO
/* prototypes for internal functions */
int POP2_sendcmd (char *cmd, int socket);
int POP2_sendHELO (char *userid, char *password, int socket);
int POP2_sendFOLD (char *folder, int socket);
int POP2_quit (int socket);
int POP2_stateGREET (int socket);
int POP2_stateNMBR (int socket);
int POP2_stateSIZE (int socket);
int POP2_stateXFER (int msgsize, int socket, int mboxfd, int topipe);
#endif


/*********************************************************************
  function:      doPOP2
  description:   retrieve messages from the specified mail server
                 using Post Office Protocol 2.

  arguments:     
    options      fully-specified options (i.e. parsed, defaults invoked,
                 etc).

  return value:  exit code from the set of PS_.* constants defined in 
                 popclient.h
  calls:         POP2_stateGREET, POP2_stateNMBR, POP2_stateSIZE,
                 POP2_stateXFER, POP2_sendcmd, POP2_sendHELO,
                 POP2_sendFOLD, POP2_quit, Socket, openuserfolder,
                 closeuserfolder, openmailpipe, closemailpipe.
  globals:       reads outlevel.
 *********************************************************************/

int doPOP2 (options)
struct optrec *options;
{
  int mboxfd;
  int socket;
  int number,msgsize,actsize;
  int status = PS_UNDEFINED;

  /* open the socket to the POP server */
  if ((socket = Socket(options->host,POP2_PORT)) < 0) {
    perror("doPOP2: socket");
    return(PS_SOCKET);
  }
    
  /* open/lock the folder if it is a user folder or stdout */
  if (options->foldertype != OF_SYSMBOX)
    if ((mboxfd = openuserfolder(options)) < 0) 
      return(PS_FOLDER);
 
  /* wait for the POP2 greeting */
  if (POP2_stateGREET(socket) != 0) {
    POP2_quit(socket);
    return(PS_PROTOCOL);
  }

  /* log the user onto the server */
  POP2_sendHELO(options->userid,options->password,socket);
  if ((number = POP2_stateNMBR(socket)) < 0) {
    POP2_quit(socket);
    return(PS_AUTHFAIL);
  }

  /* set the remote folder if selected */
  if (*options->remotefolder != 0) {
    POP2_sendFOLD(options->remotefolder,socket);
    if ((number = POP2_stateNMBR(socket)) < 0) {
      POP2_quit(socket);
      return(PS_PROTOCOL);
    }
  }

  /* tell 'em how many messages are waiting */
  if (outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%d messages in folder %s\n",number,options->remotefolder);
  else
    ;

  /* fall into a retrieve/acknowledge loop */
  if (number > 0) { 

    POP2_sendcmd("READ",socket);
    msgsize = POP2_stateSIZE(socket);
    while (msgsize > 0) {

      /* open the pipe */
      if (options->foldertype == OF_SYSMBOX)
        if ((mboxfd = openmailpipe(options)) < 0) {   
          POP2_quit(socket);
          return(PS_FOLDER);
        }

      POP2_sendcmd("RETR",socket);
      actsize = POP2_stateXFER(msgsize,socket,mboxfd,
                               options->foldertype == OF_SYSMBOX);
      if (actsize == msgsize) 
        if (options->keep)
          POP2_sendcmd("ACKS",socket);
        else
          POP2_sendcmd("ACKD",socket);
      else if (actsize >= 0) 
        POP2_sendcmd("NACK",socket);
      else {
        POP2_quit(socket);
        return(PS_SOCKET);
      }

      /* close the pipe */
      if (options->foldertype == OF_SYSMBOX)
        if (closemailpipe(mboxfd) < 0) {
          POP2_quit(socket);
          return(PS_FOLDER);
        }
    
      msgsize = POP2_stateSIZE(socket);
    }
    POP2_quit(socket);
    status = msgsize == 0 ? PS_SUCCESS : PS_PROTOCOL;
  }
  else {
    POP2_quit(socket);
    status = PS_NOMAIL;
  }

  if (options->foldertype != OF_SYSMBOX)
    closeuserfolder(mboxfd);

  return(status);
}



/*********************************************************************
  function:      POP2_sendcmd
  description:   send a command string (with no arguments) a server.
  arguments:     
    cmd          command string to send.
    socket       socket to which the server is connected.

  return value:  none.
  calls:         SockPuts.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_sendcmd (cmd,socket) 
char *cmd;
int socket;
{
  SockPuts(socket,cmd);

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> %s\n",cmd);
  else
    ;
}


/*********************************************************************
  function:      POP2_sendHELO
  description:   send the HELO command to the server.
  arguments:     
    userid       user's mailserver id.
    password     user's mailserver password.
    socket       socket to which the server is connected.

  return value:  none.
  calls:         SockPrintf.
  globals:       read outlevel.
 *********************************************************************/

int POP2_sendHELO (userid,password,socket) 
char *userid, *password;
int socket;
{
  SockPrintf(socket,"HELO %s %s\r\n",userid,password);
    

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> HELO %s password\n",userid);
  else
    ;
}


/*********************************************************************
  function:      POP2_sendFOLD
  description:   send the FOLD command to the server.
  arguments:     
    folder       name of the folder to open on the server.
    socket       socket to which the server is connected.  

  return value:  none.
  calls:         SockPrintf.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_sendFOLD (folder,socket)
char *folder;
int socket;
{
  SockPrintf(socket,"FOLD %s\r\n",folder);

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> FOLD %s\n",folder);
  else
    ;
}


/*********************************************************************
  function:      POP2_quit
  description:   send the QUIT command to the server and close 
                 the socket.

  arguments:     
    socket       socket to which the server is connected.

  return value:  none.
  calls:         SockPuts.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_quit (socket)
int socket;
{
  SockPuts(socket,"QUIT");
  close(socket);

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> QUIT\n");
  else
    ;
}


/*********************************************************************
  function:      POP2_stateGREET
  description:   process the GREET state as described in RFC 937.
  arguments:     
    socket       ...to which server is connected.

  return value:  zero if server's handling of the GREET state was 
                 correct, else non-zero (may indicate difficulty
                 at the socket).
  calls:         SockGets.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_stateGREET (socket)
int socket;
{
  char buf [POPBUFSIZE];
 
  /* read the greeting from the server */
  if (SockGets(socket, buf, sizeof(buf)) == 0) {

    /* echo the server's greeting to the user */
    if (outlevel > O_SILENT)
      fprintf(stderr,"%s\n",buf);
    else
      ;
    /* is the greeting in the correct format? */
    if (*buf == '+')
      return(0);
    else
      return(-1);
  }
  else {
    /* an error at the socket */ 
    if (outlevel > O_SILENT)
      perror("error reading socket\n");
    else
      ;
    return(-1);
  }
}


/*********************************************************************
  function:      POP2_stateNMBR
  description:   process the NMBR state as described in RFC 937.
  arguments:     
    socket       ...to which the server is connected.

  return value:  zero if the expected NMBR state action occured, else
                 non-zero.  Following HELO, a non-zero return value 
                 usually here means the user authorization at the server
                 failed.
  calls:         SockGets.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_stateNMBR (socket)
int socket;
{
  int number;
  char buf [POPBUFSIZE];

  /* read the NMBR (#ccc) message from the server */
  if (SockGets(socket, buf, sizeof(buf)) == 0) {

    /* is the message in the proper format? */
    if (*buf == '#') {
      number = atoi(buf + 1);
      if (outlevel == O_VERBOSE)
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
    else {
      number = -1;
      if (outlevel > O_SILENT) 
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
  }
  else {
    /* socket problem */
    number = -1;
    if (outlevel == O_VERBOSE) 
      perror("socket read error\n");
    else
      ;
  }
  return(number);
}


/*********************************************************************
  function:      POP2_stateSIZE
  description:   process the SIZE state as described in RFC 937.
  arguments:     
    socket       ...to which the server is connected.

  return value:  zero if the expected SIZE state action occured, else
                 non-zero (usually indicates a protocol violation).
  calls:         SockGets.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_stateSIZE (socket)
int socket;
{
  int msgsize;
  char buf [POPBUFSIZE];

  /* read the SIZE message (=ccc) from the server */
  if (SockGets(socket, buf, sizeof(buf)) == 0) 
    /* is the message in the correct format? */
    if (*buf == '=') {
      msgsize = atoi(buf + 1);
      if (outlevel == O_VERBOSE)
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
    else {
      msgsize = -1;
      if (outlevel > O_SILENT) 
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
  else {
    /* socket problem */
    msgsize = -1;
    if (outlevel == O_VERBOSE) 
      perror("socket read error\n");
    else
      ;
  }

  return(msgsize);
}


/*********************************************************************
  function:      POP2_stateXFER
  description:   process the XFER state as described in RFC 937.
  arguments:     
    msgsize      content length of the message as reported in the 
                 SIZE state.
    socket       ... to which the server is connected.
    mboxfd       open file descriptor to which the retrieved message will
                 be written.  
    topipe       true if we're writing to a the /bin/mail pipe.

  return value:  
    >= 0         actual length of the message received. 
    < 0          socket I/O problem.

  calls:         SockRead.
  globals:       reads outlevel. 
 *********************************************************************/

int POP2_stateXFER (msgsize,socket,mboxfd,topipe)
int msgsize;
int socket;
int mboxfd;
int topipe;
{
  int i,buflen,actsize;
  char buf [MSGBUFSIZE]; 
  time_t now;

  /* This keeps the retrieved message count for display purposes */
  static int msgnum = 0;  

  /* set up for status message if outlevel allows it */
  if (outlevel > O_SILENT && outlevel < O_VERBOSE) {
    fprintf(stderr,"reading message %d",++msgnum);
    /* won't do the '...' if retrieved messages are being sent to stdout */
    if (mboxfd == 1)  /* we're writing to stdout */
      fputs(".\n",stderr);
    else
      ;
  }
  else
    ;

  /* Unix mail folder format requires the Unix-syntax 'From' header.
     POP doesn't send it, so we fake it here. */
  if (!topipe) {
    now = time(NULL);
    sprintf(buf,"From POPmail %s",ctime(&now));
    if (write(mboxfd,buf,strlen(buf)) < 0) {
      perror("POP2_stateXFER: write");
      return(-1);
    }
  }

  /* read the specified message content length from the server */
  actsize = 0;
  while (msgsize > 0) {
    buflen = msgsize <= MSGBUFSIZE ? msgsize : MSGBUFSIZE;
    /* read a bufferful */ 
    if (SockRead(socket, buf, buflen) == 0) {
      /* write to folder, stripping CR chars in the process */
      for (i = 0;  i < buflen;  i++)
        if (*(buf + i) != '\r')
          if (write(mboxfd,buf + i,1) < 0) {
            perror("POP2_stateXFER: write");
            return(-1);
          }
          else
            ;  /* it was written */
        else
          ;  /* ignore CR character */
    }
    else
      return(-1);   /* socket problem */

    /* write another . for every bufferful received */
    if (outlevel > O_SILENT && outlevel < O_VERBOSE && mboxfd != 1) 
      fputc('.',stderr);
    else
      ;
    msgsize -= buflen;
    actsize += buflen;
  }

  if (!topipe) {
    /* The server may not write the extra newline required by the Unix
       mail folder format, so we write one here just in case */
    if (write(mboxfd,"\n",1) < 1) {
      perror("POP2_stateXFER: write");
      return(-1);
    }
  }
  else {
     /* the mailer might require some sort of termination string, send
        it if it is defined */
#ifdef BINMAIL_TERM
    if (write(mboxfd,BINMAIL_TERM,strlen(BINMAIL_TERM)) < 0) {
      perror("POP2_stateXFER: write");
      return(-1);
    }
#endif
  }

  /* finish up display output */
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"(%d characters of message content)\n",actsize);
  else if (outlevel > O_SILENT && mboxfd != 0)
    fputc('\n',stderr);
  else
    ;

  return(actsize);
}


config.h
/* Copyright 1993,1994 by Carl Harris, Jr.
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Carl Harris <ceharris@vt.edu>
 */


/***********************************************************************
  module:       config.h
  program:      popclient
  SCCS ID:      @(#)config.h	1.2  3/31/94
  programmer:   Carl Harris, ceharris@vt.edu
  date:         30 March 94
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  configuration defines.
 ***********************************************************************/


/***** POP defaults *****/

/* default protocol (2 = POP2, 3 = POP3) */
#define  DEF_PROTOCOL	2       


/***** Mail Delivery Agent (MDA) definitions  *****/

/* fully-qualified pathname for the MDA (usually /bin/mail) */
#define  MDA_PATH	"/bin/mail" 

/* passed as argv[0] to the MDA (usually just the basename of the mailer) */
#define  MDA_ALIAS	"mail"

/* MDA command args; $u passes user's login id (usually "-d $u") */
/* (may be undefined if the mailer doesn't require arguments (?!)) */
#define  MDA_ARGS	"-d $u"

/* MDA command arg count -- must be greater than or equal to the number 
   of space-delimited arguments defined in MDA_ARGS. */
#ifdef MDA_ARGS
#define MDA_ARGCOUNT	8
#endif
   
/* MDA input message terminator (usually not required) */
/* (may be undefined if the mailer doesn't require it) */
/* #define MDA_TERM  ".\n" */



popclient.c
/* Copyright 1993,1994 by Carl Harris, Jr.
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Carl Harris <ceharris@vt.edu>
 */

 
/***********************************************************************
  module:       popclient.c
  program:      popclient
  SCCS ID:      @(#)popclient.c	2.5  4/1/94
  programmer:   Carl Harris, ceharris@vt.edu
  date:         29 December 1993
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  main driver module for the popclient package.  contains
                code for command-line parsing and invocation of the
                appropriate POP driver routine.
 ***********************************************************************/
 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <pwd.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/file.h>
#include <sys/wait.h>
#include <sys/stat.h>

#include "config.h"
#include "popclient.h"

/* release info */
#define         COMPILE_SCCS	"2.21 [2.5] (4/1/94)"

/* requisite, venerable SCCS ID string */
static char sccs_id [] = "@(#)popclient.c	2.5\t4/1/94";

#ifndef NO_PROTO
/* prototypes for internal functions */
int showoptions (struct optrec *options);
int parseMDAargs (struct optrec *options);
int parsecmdline (int argc, char **argv, struct optrec *options);
int setdefaults (struct optrec *options);
int showversioninfo (void);
int clearbiffbit (void);
int restorebiffbit (void);
#endif

/* Controls the detail of status/progress messages written to stderr */
int outlevel;      /* see the O_.* constants in popclient.h */
int biffwas = -1;  /* previous chmod bits on the controlling tty */

/* args for the MDA, parsed out in the usual fashion by parseMDAargs() */
#ifdef MDA_ARGS
char *mda_argv [MDA_ARGCOUNT + 2];
#else
char *mda_argv [2];
#endif


/*********************************************************************
  function:      main
  description:   main driver routine 
  arguments:     
    argc         argument count as passed by runtime startup code.
    argv         argument strings as passed by runtime startup code.

  return value:  an exit status code for the shell -- see the 
                 PS_.* constants defined above.
  calls:         parsecmdline, setdefaults, openuserfolder, doPOP2.
  globals:       none.
 *********************************************************************/

main (argc,argv)
int argc;
char **argv;
{ 
  int mboxfd;
  struct optrec options;
  int popstatus;
  int cleanup();

  signal(SIGINT,cleanup);
  clearbiffbit();

  if (parsecmdline(argc,argv,&options) == 0)
    if (!options.versioninfo)
      if (setdefaults(&options) == 0) {
        parseMDAargs(&options);
        if (options.whichpop == 3) 
          popstatus = doPOP3(&options);
        else
          popstatus = doPOP2(&options);
      } 
      else
        popstatus = PS_UNDEFINED;
    else
      showversioninfo();
  else
    popstatus = PS_SYNTAX;

  restorebiffbit();
  exit(popstatus);
}
    

 
/*********************************************************************
  function:      parsecmdline
  description:   parse/validate the command line options.
  arguments:
    argc         argument count.
    argv         argument strings.
    options      pointer to a struct optrec to receive the parsed 
                 options.

  return value:  zero if the command line was syntactically correct,
                 else non-zero (an error message has been written to stderr).  
  calls:         none.  
  globals:       none.  
 *********************************************************************/

int parsecmdline (argc,argv,options)
int argc;
char **argv;
struct optrec *options;
{
  int c,i;
  int fflag = 0;     /* TRUE when -o or -c has been specified */
  int errflag = 0;   /* TRUE when a syntax error is detected */
  extern int optind,opterr;     /* defined in getopt(2) */
  extern char *optarg;          /* defined in getopt(2) */

  bzero(options,sizeof(struct optrec));    /* start clean */
  while (!errflag && (c = getopt(argc,argv,"23Vkvscu:p:f:o:")) != EOF) {
    switch (c) {
      case '2':
        options->whichpop = 2;
        break;
      case '3':
        options->whichpop = 3;
        break;
      case 'V':
        options->versioninfo = !0;
        break;
      case 'k':
        options->keep = !0;
        break;
      case 'v':
        options->verbose = !0;
        break;
      case 's':
        options->silent = !0;
        break;
      case 'c':
        if (fflag)
          errflag++;
        else {
          fflag++;
          options->foldertype = OF_STDOUT;
        }
        break;
      case 'u':
        strcpy(options->userid,optarg);
        break;
      case 'p':
        strcpy(options->password,optarg);
        break;
      case 'o':
        if (fflag) 
          errflag++;
        else {
          fflag++;
          options->foldertype = OF_USERMBOX;
          strcpy(options->userfolder,optarg);
        }
        break;
      case 'f':
        strcpy(options->remotefolder,optarg);
        break;
      default:
        errflag++;
    }
  }

  if (!options->versioninfo) {
    /* get host argument -- error if not present */
    if (optind < argc)
      strcpy(options->host,argv[optind++]);
    else
      errflag++;

    /* error if more unparsed arguments are present */
    if (optind < argc)
      errflag++;
    else
      ;
  }
  else 
    ;


  /* squawk if syntax errors were detected */
  if (errflag) {
    fputs("usage:  popclient [-2|-3] [-Vksv] [-u server-userid] [-p server-password]\n",
          stderr);
    fputs("                  [-f remote-folder] [-c | -o local-folder]  host\n",
          stderr);
  }
  else
    ;

  /* clean up the command line in case -p was used */
  for (i = 1; i < argc; i++)
    argv[i] = (char *) 0;

  return(errflag);
}
         

/*********************************************************************
  function:      setdefaults
  description:   set reasonable default values for unspecified options.
  arguments:     
    options      option values parsed from the command-line; unspeci-
                 fied options must be filled with zero.

  return value:  zero if defaults were successfully set, else non-zero
                 (indicates a problem reading /etc/passwd).
  calls:         none.
  globals:       writes outlevel.
 *********************************************************************/

int setdefaults (options)
struct optrec *options;
{
  int uid;
  struct passwd *pw;
  char *mailvar;

  if ((pw = getpwuid(uid = getuid())) == NULL) {
    fprintf(stderr,"No passwd entry for uid %d\n",uid);
    return(-1);
  }
  /* save the login name for delivery use */
  strcpy(options->loginid,pw->pw_name);

  if (options->whichpop == 0)
    options->whichpop = DEF_PROTOCOL;
  else
    ; /* -2 or -3 was specified */

  if (*(options->userid) == '\0') 
    strcpy(options->userid,pw->pw_name);
  else
    ;  /* -u was specified */

  if (options->foldertype == 0) 
    options->foldertype = OF_SYSMBOX;
  else
    ;  /* -o or -c was specified */
    
  if (*(options->password) == '\0')  
    strcpy(options->password,(char *) getpass("Enter mailserver password: "));
  else
    ;  /* -p was specified */

  if (options->verbose)
    outlevel = O_VERBOSE;
  else if (options->silent)
    outlevel = O_SILENT;
  else
    outlevel = O_NORMAL;

  return(0);
}



/*********************************************************************
  function:      showversioninfo
  description:   display program release and compiler info
  arguments:     none.
  return value:  none.
  calls:         none.
  globals:       none.
 *********************************************************************/

int showversioninfo()
{
  printf("popclient release %s\n",COMPILE_SCCS);
}



/*********************************************************************
  function:      openuserfolder
  description:   open the file to which the retrieved messages will
                 be appended.  Do NOT call when options->foldertype
                 is OF_SYSMBOX.

  arguments:     
    options      fully-determined options (i.e. parsed, defaults invoked,
                 etc).

  return value:  file descriptor for the open file, else -1.
  calls:         none.
  globals:       none.
 *********************************************************************/

int openuserfolder (options)
struct optrec *options;
{
  int fd;

  if (options->foldertype == OF_STDOUT)
    return(1);
  else    /* options->foldertype == OF_USERMBOX */
    if ((fd = open(options->userfolder,O_CREAT|O_WRONLY|O_APPEND,0600)) >= 0) {
      return(fd);
    }
    else {
      perror("popclient: openuserfolder: open()");
      return(-1);
    }
  
}



/*********************************************************************
  function:      openmailpipe
  description:   open a one-way pipe to the mail delivery agent.
  arguments:     
    options      fully-determined options (i.e. parsed, defaults invoked,
                 etc).

  return value:  open file descriptor for the pipe or -1.
  calls:         none.
  globals:       reads mda_argv.
 *********************************************************************/

int openmailpipe (options)
struct optrec *options;
{
  int pipefd [2];
  int childpid;
  char binmailargs [80];

  if (pipe(pipefd) < 0) {
    perror("popclient: openmailpipe: pipe");
    return(-1);
  }
  if ((childpid = fork()) < 0) {
    perror("popclient: openmailpipe: fork");
    return(-1);
  }
  else if (childpid == 0) {

    /* in child process space */
    close(pipefd[1]);  /* close the 'write' end of the pipe */
    close(0);          /* get rid of inherited stdin */
    if (dup(pipefd[0]) != 0) {
      fputs("popclient: openmailpipe: dup() failed\n",stderr);
      exit(1);
    }

    execv(MDA_PATH,mda_argv);

    /* if we got here, an error occurred */
    perror("popclient: openmailpipe: exec");
    return(-1);

  }

  /* in the parent process space */
  close(pipefd[0]);  /* close the 'read' end of the pipe */
  return(pipefd[1]);
}



/*********************************************************************
  function:      closeuserfolder
  description:   close the user-specified mail folder.
  arguments:     
    fd           mail folder descriptor.

  return value:  zero if success else -1.
  calls:         none.
  globals:       none.
 *********************************************************************/

int closeuserfolder(fd)
int fd;
{
  int err;

  if (fd != 1) {   /* not stdout */
    err = close(fd);
  }   
  else
    err = 0;
  
  if (err)
    perror("popclient: closeuserfolder: close");

  return(err);
}



/*********************************************************************
  function:      closemailpipe
  description:   close pipe to the mail delivery agent.
  arguments:     
    options      fully-determined options record
    fd           pipe descriptor.

  return value:  0 if success, else -1.
  calls:         none.
  globals:       none.
 *********************************************************************/

int closemailpipe (fd)
int fd;
{
  int err;
  int childpid;

  err = close(fd);
  childpid = wait((union wait *) 0);
  if (err)
    perror("popclient: closemailpipe: close");

  return(err);
}



/*********************************************************************
  function:      parseMDAargs
  description:   parse the argument string given in MDA_ARGS into
                 a regular *argv[] array.
  arguments:
    options      fully-determined options record pointer.

  return value:  none.
  calls:         none.
  globals:       writes mda_argv.
 *********************************************************************/

int parseMDAargs (options)
struct optrec *options;
{
  int argi;
  char *argp;

  /* first put the MDA alias in as argv[0] */
  mda_argv[0] = MDA_ALIAS;
  
#ifdef MDA_ARGS

  /* make a writeable copy of MDA_ARGS */
  argp = strcpy((char *) malloc(strlen(MDA_ARGS)+1), MDA_ARGS);
  
  while (*argp != '\0' && isspace(*argp))	/* skip null first arg */
    argp++;					

  /* now punch nulls into the delimiting whitespace in the args */
  for (argi = 1;  
       *argp != '\0';
       argi++) {

    mda_argv[argi] = argp;     /* store pointer to this argument */

    /* find end of this argument */
    while (!(*argp == '\0' || isspace(*argp)))
      argp++;

    /* punch in a null terminator */
    if (*argp != '\0')
      *(argp++) = '\0';  
 
    /* check for macros */
    if (strcmp(mda_argv[argi],"$u") == 0)
      mda_argv[argi] = 
        strcpy((char *) malloc(strlen(options->loginid)+1),options->loginid);
    else
      ;  /* no macros to expand */

  }
  mda_argv[argi] = (char *) 0;

#else 

  mda_argv[1] = (char *) 0;

#endif

}



/*********************************************************************
  function:      clearbiffbit
  description:   clear the owner execute bit on the controlling tty, so
                 that biff messages do not appear on the terminal with
                 the program output.

  arguments:     none.
  return value:  none.
  calls:         none.
  globals:       writes biffwas
 *********************************************************************/

int clearbiffbit ()
{
  struct stat buf;
 
  if (fstat(0,&buf) < 0) {
    perror("popclient: clearbiffbit: fstat");  /* fuss but don't quit */
    return;
  }

  if (buf.st_mode & S_IFCHR) {
    buf.st_mode &= 0777;        /* isolate the device permissions */
    biffwas = buf.st_mode;	/* save for restore */
    buf.st_mode &= ~S_IEXEC;    /* clear the execute bit */
    if (fchmod(0,buf.st_mode) < 0) {
      perror("popclient: clearbiffbit: fchmod");  /* fuss but don't quit */
      return;
    }
  }
}




/*********************************************************************
  function:      restorebiffbit
  description:   restore the previous state of the owner execute bit
                 on the controlling tty.
  arguments:     none.
  return value:  none.
  calls:         none.
  globals:       reads biffwas
 *********************************************************************/

int restorebiffbit ()
{
  if (biffwas == -1)
    return;    /* not a tty */
  else {
    if (fchmod(0,biffwas) < 0) 
      perror("popclient: restorebiffbit: fchmod");   /* fuss but don't quit */
  }
}



/*********************************************************************
  function:      cleanup
  description:   signal catcher.
  arguments:     none.
  return value:  none.
  calls:         restorebiff
  globals:       reads biffwas.
 *********************************************************************/
    
int cleanup ()
{
  if (biffwas != -1) 
    restorebiffbit();
}


